### PR TITLE
Feature/neo 623 map serializer as arrays

### PR DESCRIPTION
**Pull request by mpluecker**

### CODE REVIEW

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  private final Map<JsonNode, JsonNode> children;
+
+  public static JsonNode deepConvertNode(JsonNode node) {
+    if (node.isObject()) {
+      if (node.size() == 1) {
+        Entry<String, JsonNode> arrayMapField = node.fields().next();
+        if (arrayMapField.getKey().equals(ArrayMapSerializer.ARRAY_MAP) && arrayMapField.getValue().isArray()) {
+          return convertToArrayNode(node, arrayMapField);
+        }
+      }
```
Comment (sschmittBt) [inline src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java:49]: Can be replaced by "isArrayMapNode"

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+  /**
+   * Method that can be called to serialize this node and
+   * all of its descendants using specified JSON generator.
+   */
+  @Override
+  public void serialize(JsonGenerator g, SerializerProvider provider)
+      throws IOException {
+    g.writeStartObject();
+    innserSerialize(g, provider);
```
Comment (sschmittBt) [inline src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java]: typo, "inner"

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    node.set(ArrayMapSerializer.ARRAY_MAP, childrenNode);
+    return node;
+  }
+
+  @Override
+  public void serializeWithType(JsonGenerator g, SerializerProvider provider,
+      TypeSerializer typeSer)
+      throws IOException {
+    @SuppressWarnings("deprecation")
+    boolean trimEmptyArray = (provider != null) &&
```
Comment (sschmittBt) [inline src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java]: What do you need this boolean for?

File: src/main/java/de/braintags/vertx/util/json/JsonDiff.java [hunk 4]
```diff
@@ -55,18 +65,31 @@ public static JsonNode getDiff(JsonNode base, JsonNode data) {
...
+      case OBJECT:
+        if (!arrayMapsConverted && ArrayMapNode.isArrayMapNode(base)) {
+          ArrayMapNode arrayBase = (ArrayMapNode) ArrayMapNode.deepConvertNode(base);
+          ArrayMapNode arrayData = (ArrayMapNode) ArrayMapNode.deepConvertNode(data);
+          return arrayMapDiff(arrayBase, arrayData, nodeFactory, true);
+        } else {
+          return objectDiff((ObjectNode) base, (ObjectNode) data, nodeFactory, arrayMapsConverted);
+        }
+      case POJO:
+        return arrayMapDiff((ArrayMapNode) base, (ArrayMapNode) data, nodeFactory, arrayMapsConverted);
```
Comment (sschmittBt) [inline src/main/java/de/braintags/vertx/util/json/JsonDiff.java]: How can you assume that every POJO node is already an ArrayMapNode?

File: src/main/java/de/braintags/vertx/util/json/deserializers/ArrayMapDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+                value = deserialize(p, ctxt, valueDes, typeDeser);
+                if (useObjectId) {
+                  referringAccumulator.put(key, value);
+                } else {
+                  result.put(key, value);
+                }
+              } catch (UnresolvedForwardReference reference) {
+                handleUnresolvedReference(p, referringAccumulator, key, reference);
+              } catch (Exception e) {
+                wrapAndThrow(e, result, key.toString());
```
Comment (sschmittBt) [inline src/main/java/de/braintags/vertx/util/json/deserializers/ArrayMapDeserializer.java]: key might be null if value came before the key

File: src/test/java/de/braintags/vertx/util/json/TArrayMap.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+import de.braintags.vertx.util.json.deserializers.ArrayMapSerializer;
+import de.braintags.vertx.util.json.deserializers.ComplexKey;
+import io.vertx.core.json.Json;
+
+/**
+ * Unit test for {@link MapAsArraySerializer}
+ * 
+ *
+ */
+public class TArrayMap {
```
Comment (sschmittBt) [inline src/test/java/de/braintags/vertx/util/json/TArrayMap.java:29]: What about array maps in array maps? Or is it unsupported?

**Review by sschmittBt — CHANGES_REQUESTED** [review]

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java (hunk 1 above)
Comment (mpluecker) [reply src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java:49]: this is intended, using "isArrayMapNode" would be slower

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java (hunk 3 above)
Comment (mpluecker) [reply src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java]: nothing .. done

File: src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java (hunk 2 above)
Comment (mpluecker) [reply src/main/java/de/braintags/vertx/util/json/ArrayMapNode.java]: done

File: src/test/java/de/braintags/vertx/util/json/TArrayMap.java (hunk 6 above)
Comment (mpluecker) [reply src/test/java/de/braintags/vertx/util/json/TArrayMap.java:29]: added support and test for that

File: src/main/java/de/braintags/vertx/util/json/deserializers/ArrayMapDeserializer.java (hunk 5 above)
Comment (mpluecker) [reply src/main/java/de/braintags/vertx/util/json/deserializers/ArrayMapDeserializer.java]: done

**Review by sschmittBt — APPROVED** [review]

**Pull request merged**